### DISCUSSION TITLE
Anonymous type definitions are not allowed

### DISCUSSION THREAD

**Discussion by goto40** [RemedyIT/ridl#138] [Q&A] [answered]
I have an IDL file like this:
```
// fastddsgen -replace HelloWorld.idl

enum ColorT
{
    RED,
    GREEN,
    BLUE
};

struct HelloWorld
{
    @key unsigned long id;
    unsigned long index;
    string message;
    ColorT color;
    sequence<ColorT> colors;
    long values[3];
};
```

I get `IDL::ParseError: Anonymous type definitions are not allowed!` for the ridl compiler. I can get rid of by introducing typedefs. Is there a way to enable this anonymous type definitions?

**Comment by jwillemsen** [comment, marked as answer]
The short answer is no, RIDL doesn't support anonymous types, you *must* use a typedef in your IDL.

The longer answer is that IDL3 deprecated anonymous types because it is problematic in some language mappings, see [IDL3.5](https://www.omg.org/spec/IDL/3.5/PDF) chapter 5.11.6. Anonymous types are problematic in some language mappings, for example in the IDL to C++11 mapping. In that language mapping we declare a `IDL::traits<>` specialization for your types, but we can't do that for your `sequence<ColorT>`. Also a lot of questions arise when we use distinct types, what when you use the same anonymous type in different modules, is that the same type or not? Because of this the CORBA vendors in the past deprecated anonymous types.

At some point the DDS vendors decided that they do want to support anonymous types, they only use a subset of IDL. With the definition of IDL4 anonymous types where added back as part of an optional building block. They completely ignored the fact that this is problematic, but their scope and usage of IDL is very limited.

Because the available open source backends of RIDL don't support anonymous types we added the error checking for anonymous types to the RIDL frontend. At some point we discussed migrating those checks to the various backends but that is very very low on the priority list. The easiest and best solution is to introduce typedefs, that makes your IDL more clean and it will work with any language mapping.
